File: frontend/src/types/locationReopen.ts
// Types for `location_reopen_request` — the only path that moves a
// `closed_pending_reopen` location back to `active`. Matches
// backend/app/schemas/location_reopen.py, deliberately shaped like
// @/types/claim.ts's ClaimResponse/ClaimQueueItem/Approve/Reject — same
// admin-review pattern, same field names where the concept is the same.

export type ReopenRequestStatus = "pending_review" | "approved" | "rejected";

/** Body for POST /locations/{id}/reopen-requests. */
export interface CreateReopenRequestInput {
  notes?: string | null;
}

/** Response for POST /locations/{id}/reopen-requests and GET /location-reopen-requests/{id}. */
export interface ReopenRequestResponse {
  request_id: number;
  location_id: number;
  status: ReopenRequestStatus;
  notes: string | null;
  submitted_at: string;
  reviewed_at: string | null;
  reviewer_notes: string | null;
}

/** Row of GET /location-reopen-requests (admin queue). */
export interface ReopenRequestQueueItem {
  request_id: number;
  location_id: number;
  brand_id: number;
  brand_name: string;
  brand_slug: string;
  location_address: string;
  requested_by_user_id: string;
  /** null when no owner_account row matches the requester. */
  requester_email: string | null;
  notes: string | null;
  status: ReopenRequestStatus;
  submitted_at: string;
  reviewed_by: string | null;
  reviewed_at: string | null;
  reviewer_notes: string | null;
}

/** Body for POST /location-reopen-requests/{id}/approve. */
export interface ApproveReopenRequestInput {
  reviewer_notes?: string;
}

/** Body for POST /location-reopen-requests/{id}/reject. reviewer_notes is required. */
export interface RejectReopenRequestInput {
  reviewer_notes: string;
}

File: backend/app/schemas/location_reopen.py
"""location_reopen_request — see app/models/location_reopen_request.py and
docs/API_CONTRACTS.md "Location reopen requests
(`/locations/{id}/reopen-requests`, `/location-reopen-requests`)".

Shape deliberately mirrors app/schemas/claim.py (ClaimCreate/ClaimOut/
ClaimQueueItem/ClaimApproveRequest/ClaimRejectRequest) — same admin-review
pattern, same field names where the concept is the same (`reviewer_notes`,
`status`, `submitted_at`, `reviewed_at`).
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

NOTES_MAX_LENGTH = 2000

ReopenRequestStatus = Literal["pending_review", "approved", "rejected"]


class ReopenRequestCreate(BaseModel):
    """Body for `POST /locations/{id}/reopen-requests` — `location_id`
    comes from the path (same pattern as `POST /locations/{id}/managers`'s
    `AssignManagerRequest`), so only the owner's optional notes are here."""

    notes: str | None = Field(default=None, max_length=NOTES_MAX_LENGTH)


class ReopenRequestOut(BaseModel):
    """Response for `POST /locations/{id}/reopen-requests` and
    `GET /location-reopen-requests/{id}`."""

    request_id: int
    location_id: int
    status: ReopenRequestStatus
    notes: str | None = None
    submitted_at: datetime
    reviewed_at: datetime | None = None
    reviewer_notes: str | None = None


class ReopenRequestQueueItem(BaseModel):
    """Admin-facing row (`GET /location-reopen-requests`) — same shape
    family as `ClaimQueueItem`, with the brand/location context a bare
    `location_id` wouldn't give the admin reviewer."""

    request_id: int
    location_id: int
    brand_id: int
    brand_name: str
    brand_slug: str
    location_address: str
    requested_by_user_id: str
    # None when no owner_account row matches the requester — same
    # nullability reasoning as ClaimQueueItem.claimant_email.
    requester_email: str | None = None
    notes: str | None = None
    status: ReopenRequestStatus
    submitted_at: datetime
    reviewed_by: str | None = None
    reviewed_at: datetime | None = None
    reviewer_notes: str | None = None


class ReopenRequestListResponse(BaseModel):
    results: list[ReopenRequestQueueItem]
    page: int
    page_size: int
    total: int


class ReopenRequestApproveRequest(BaseModel):
    reviewer_notes: str | None = Field(default=None, max_length=NOTES_MAX_LENGTH)


class ReopenRequestRejectRequest(BaseModel):
    reviewer_notes: str = Field(min_length=1, max_length=NOTES_MAX_LENGTH)
